README.md
netperf-wrapper
===============

Python wrapper to run multiple simultaneous netperf/iperf/ping instances and
aggregate the results.

Tests are specified as .ini config files, and various parsers for tool output
are supplied. At the moment, parsers for netperf in -D mode, iperf in csv mode
and ping/ping6 in -D mode are supplied, as well as a generic parser for commands
that just outputs a single number.

Several tests can be run in parallel and, provided they output timestamped
values, (which netperf ping and iperf do, the latter with a small patch,
available in the misc/ directory), the test data points can be aligned with each
other in time, interpolating differences between the actual measurement points.
This makes it possible to graph (e.g.) ping times before, during and after a
link is loaded.

An alternative run mode is running several iterated tests (which each output one
data point, e.g. netperf tests not in -D mode), and outputting the results of
these several runs.

Output is either simple Python pretty-printed output (which can be eval'ed back
into a Python data structure), org mode tables, or plots. The latter requires a
functioning matplotlib installation. You specify which output format with -f {org,plot} and for plot formats you specify -o output.{png,ps,pdf,svg}.

Documentation is relatively sparse at the moment, but try having a look at the
.ini files in the tests directory and running the main script
(netperf-wrapper.py) with -h.

## Saving test data for later plotting ##

Reading back in test data is now supported. First, run the tests with the
'pprint' formatter and save the output in a file (i.e. `python2
netperf-wrapper.py -f pprint -o test.data <test>`), then read it back in using
the -i parameter. This bypasses the actual running of the tests, but still
requires a test config file. So for example, the data can be plotted by using
`python2 netperf-wrapper.py -f plot -i test.data <test>` (with the same test
name as was used to generate the data).

Note that this functionality relies on the data to be eval()'ed by the python
interpreter, so don't under any circumstances use untrusted input data, as it
can in principle do arbitrary bad things.


CHANGES
2014-05-02: Add --subplot-combine option to combine multiple data sets into
            separate subplots rather than plotting them onto the same plot.

2014-05-02: Manage plot series colours to make them consistent when combining
            plots, and instead cycle line styles and markers to differentiate
            result sets in a combined plot.

2014-05-02: Fix plot scaling on box plots.

2014-05-02: Fix several bugs in the resizing of plots to make room for the
            legend (in interactive mode).

2014-04-30: Fix parsing of netperf output to not include spurious results at the end
            of UDP measurement flows giving a huge RTT because the measurement flow
            was lost during the test.

2014-04-30: Fix SIGINT behaviour so netperf-wrapper exits by SIGINT'ing itself after
            having cleaned up runners etc.

2014-04-30: Add batch run mode to automatically run several parameterised tests.

2014-04-09: On GUI startup, by default search for a running instance and pass
            the file name to that rather than starting a new instance.

2014-04-09: Add .desktop file and mime info for data files.

2014-04-09: Add toggle to invert y-axis on latency measurements.

2014-04-04: Add Q-Q plot type.

2014-03-24: Allow tests to specify axis labels explicitly. Use this to write
            something better than 'ms' on the latency axes.

2014-03-24: Add box plot type.

2014-03-23: Disable hostname lookup when loading data files or the GUI; should
            fix issue with hostname lookup failing when a datafile with an
            unreachable host is loaded.

2014-03-23: Add GUI for interactive browsing of results.

2014-03-20: rc-file should specify hosts rather than host.

2014-03-19: Record path towards machine running test in INGRESS_INFO of remote
            metadata object.

2014-03-19: Capture IP address info per-interface.

2014-03-16: Scale x axis to include all captures data points, even if they go beyond
            the configured test duration.

2014-03-16: Run ping with -n.

2014-03-05: Add remote metadata support.

2014-03-04: More extended metadata.

2014-03-01: Support an rcfile and supply an example file.

2014-03-01: Add metadata output formatter to extract metadata from stored data files.

2014-03-01: Add collection of several extended metadata items.

2014-01-22: Initial provisions for optionally gathering extended metadata on test runs.

2014-01-19: Add detection of supported netperf options and support passing -e to
            netperf for UDP_RR socket timeout (adds --socket-timeout parameter).

2014-01-18: Add rrul_be test using exclusively BE traffic.

2013-12-25: Don't crash on empty values in CDF plots.

2013-11-25: Release v0.6.1.

2013-11-25: Fix module rewrite in setup.py and numpy version output in --version.

2013-11-25: Release v0.6.0.

2013-10-31: Adjust spacing for title and annotation to work better with exported plots.

2013-09-27: Add description of output formats to man page.

2013-09-27: Add the --note argument to store arbitrary notes in the JSON metadata.

2013-09-27: Print custom title on separate line on plots.

2013-09-26: Add man page.

2013-09-25: Allow passing of data filenames and host names as unqualified arguments
            and attempt to guess what each argument is.

2013-09-25: Add matplotlib version to --version output.

2013-04-20: Manual removal of temporary files to get around issues on cygwin.

2013-04-18: Print Python version along with netperf version on -V.

2013-04-18: Fix python3.1 compatibility bug in gzip workaround.

2013-04-15: Release v0.5.6.

2013-04-15: Add rrul_noclassification test.

2013-04-15: Release v0.5.5.

2013-04-15: Add work-around for python2.6 version of GzipFile.

2013-04-15: Add option to disable automatic log-scaling of graphs.

2013-03-27: Add -z option to start y-axis at zero.

2013-03-27: Don't use suptitle() to draw annotation at the bottom of the plots (makes
            the annotation work in matplotlib 1.2.1).

2013-03-27: Release v0.5.3.

2013-03-27: Fix install bug in setup.py when creating packages.

2013-03-27: Release v0.5.2.

2013-03-26: Actually get the return code of a subprocess.

2013-03-26: Fix bug on aborting tests due to missing kill() stub for computing runners.

2013-03-26: Rework the way access checks are done for output files.

2013-03-18: Release v0.5.1.

2013-03-18: Compatibility fix for Python 3.1.

2013-03-18: Release v0.5.0.

2013-03-18: Do not crash if killing subprocess fails because it is already killed.

2013-03-18: Ensure compatibility with a wider range of python versions.

2013-03-18: Do not fail horribly on write permission errors.

2013-03-11: Preforking runner implementation (hopefully fixes issues on FreeBSD).

2013-03-11: Be smarter about picking a `ping` binary, trying first fping
            then falling back to straight ping.

2013-02-20: Don't crash on missing data, just print a warning.

2013-02-20: Workaround for bad exit codes on bsd.

2013-02-14: Release v0.4.2.

2013-01-25: Encode title in data set filename if set.

2013-01-22: Fix possible divide by zero error on certain data values.

2013-01-22: Get rid of the colons in the time stamps used for file names (they
            give rise to escaping headaches).

2012-12-20: Release v0.4.1.

2012-12-20: Fix cutoff logic when end cutoff value is 0.

2012-12-20: If there are missing data values when doing a CDF plot, the plot
            should cut off at a probability < 1; i.e. use the total size of the
            data (including null data points) as data set size when calculating
            probability.

2012-12-17: Release v0.4.0.

2012-12-15: Add stats formatter (-f stats) to output statistics on results data sets.

2012-12-13: Fix bug that caused smoothing to add new data points at the ends of
            series.

2012-12-13: Fix bug that caused primary axis to place ticks on the right hand
            side in dual-axis plots.

2012-12-13: Release v0.3.0.


formatters.py
## -*- coding: utf-8 -*-
##
## formatters.py
##
## Author:   Toke Høiland-Jørgensen (toke@toke.dk)
## Date:     16 oktober 2012
## Copyright (c) 2012, Toke Høiland-Jørgensen
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see <http://www.gnu.org/licenses/>.

import pprint, sys, csv, math

from settings import settings
from util import cum_prob, frange

PLOT_KWARGS = (
    'alpha',
    'antialiased',
    'color',
    'dash_capstyle',
    'dash_joinstyle',
    'drawstyle',
    'fillstyle',
    'label',
    'linestyle',
    'linewidth',
    'lod',
    'marker',
    'markeredgecolor',
    'markeredgewidth',
    'markerfacecolor',
    'markerfacecoloralt',
    'markersize',
    'markevery',
    'pickradius',
    'solid_capstyle',
    'solid_joinstyle',
    'visible',
    'zorder'
    )

class Formatter(object):

    def __init__(self, output):
        if isinstance(output, basestring):
            if output == "-":
                self.output = sys.stdout
            else:
                self.output = open(output, "w")
        else:
            self.output = output

    def format(self, results):
        sys.stderr.write(u"No output formatter selected.\nTest data is in %s (use with -i to format).\n" % results.dump_file)

DefaultFormatter = Formatter


class OrgTableFormatter(Formatter):
    """Format the output for an Org mode table. The formatter is pretty crude
    and does not align the table properly, but it should be sufficient to create
    something that Org mode can correctly realign."""

    def format(self, results):
        name = results.meta("NAME")

        if not results:
            self.output.write(unicode(name) + u" -- empty\n")
        keys = settings.DATA_SETS.keys()
        header_row = [name] + keys
        self.output.write(u"| " + u" | ".join(header_row) + u" |\n")
        self.output.write(u"|-" + u"-+-".join([u"-"*len(i) for i in header_row]) + u"-|\n")

        def format_item(item):
            if isinstance(item, float):
                return "%.2f" % item
            return unicode(item)

        for row in results.zipped(keys):
            self.output.write(u"| ")
            self.output.write(u" | ".join(map(format_item, row)))
            self.output.write(u" |\n")



class CsvFormatter(Formatter):
    """Format the output as csv."""

    def format(self, results):
        name = results.meta("NAME")

        if not results:
            return

        writer = csv.writer(self.output)
        keys = settings.DATA_SETS.keys()
        header_row = [name] + keys
        writer.writerow(header_row)

        def format_item(item):
            if item is None:
                return ""
            return unicode(item)

        for row in results.zipped(keys):
            writer.writerow(map(format_item, row))

class PlotFormatter(Formatter):

    def __init__(self, output):
        self.output = output
        try:
            import matplotlib, numpy
            # If saving to file, try our best to set a proper backend for
            # matplotlib according to the output file name. This helps with
            # running matplotlib without an X server.
            if output != "-":
                if output.endswith('.svg') or output.endswith('.svgz'):
                    matplotlib.use('svg')
                elif output.endswith('.ps') or output.endswith('.eps'):
                    matplotlib.use('ps')
                elif output.endswith('.pdf'):
                    matplotlib.use('pdf')
                elif output.endswith('.png'):
                    matplotlib.use('agg')
                else:
                    raise RuntimeError("Unrecognised file format for output '%s'" % output)
            import matplotlib.pyplot as plt
            self.plt = plt
            self.np = numpy
            self._init_plots()
        except ImportError:
            raise RuntimeError(u"Unable to plot -- matplotlib is missing! Please install it if you want plots.")


    def _load_plotconfig(self, plot):
        if not plot in settings.PLOTS:
            raise RuntimeError(u"Unable to find plot configuration '%s'" % plot)
        config = settings.PLOTS[plot]
        if 'parent' in config:
            parent_config = settings.PLOTS[config['parent']]
            parent_config.update(config)
            return parent_config
        return config

    def _init_plots(self):
        self.config = self._load_plotconfig(settings.PLOT)
        getattr(self, '_init_%s_plot' % self.config['type'])()

    def _init_timeseries_plot(self, config=None, axis=None):
        if axis is None:
            axis = self.plt.gca()
        if config is None:
            config = self.config

        if 'dual_axes' in config and config['dual_axes']:
            second_axis = self.plt.axes(axis.get_position(), sharex=axis, frameon=False)
            second_axis.yaxis.tick_right()
            second_axis.yaxis.set_label_position('right')
            second_axis.yaxis.set_offset_position('right')
            second_axis.xaxis.set_visible(False)
            config['axes'] = [axis,second_axis]
        else:
            config['axes'] = [axis]


        unit = [None]*len(config['axes'])
        for s in config['series']:
            if 'axis' in s and s['axis'] == 2:
                a = 1
            else:
                a = 0
            s_unit = settings.DATA_SETS[s['data']]['units']
            if unit[a] is not None and s_unit != unit[a]:
                raise RuntimeError("Plot axis unit mismatch: %s/%s" % (unit[a], s_unit))
            unit[a] = s_unit

        axis.set_xlabel('Time')
        for i,u in enumerate(unit):
            config['axes'][i].set_ylabel(unit[i])

    def _init_cdf_plot(self, config=None, axis=None):
        if axis is None:
            axis = self.plt.gca()
        if config is None:
            config = self.config

        unit = None
        for s in config['series']:
            s_unit = settings.DATA_SETS[s['data']]['units']
            if unit is not None and s_unit != unit:
                raise RuntimeError("Plot axis unit mismatch: %s/%s" % (unit, s_unit))
            unit = s_unit

        axis.set_xlabel(unit)
        axis.set_ylabel('Cumulative probability')
        axis.set_ylim(0,1)
        config['axes'] = [axis]


    def _init_meta_plot(self):
        self.configs = []
        for i,subplot in enumerate(self.config['subplots']):
            axis = self.plt.subplot(len(self.config['subplots']),1,i+1, sharex=self.plt.gca())
            config = self._load_plotconfig(subplot)
            self.configs.append(config)
            getattr(self, '_init_%s_plot' % config['type'])(config=config, axis=axis)
            if i < len(self.config['subplots'])-1:
                axis.set_xlabel("")


    def _do_timeseries_plot(self, results, config=None, axis=None):
        if axis is None:
            axis = self.plt.gca()
        if config is None:
            config = self.config

        axis.set_xlim(0, settings.TOTAL_LENGTH)
        data = []
        for i in range(len(config['axes'])):
            data.append([])

        for s in config['series']:
            if 'smoothing' in s:
                smooth=s['smoothing']
            else:
                smooth = False
            kwargs = {}
            for k in PLOT_KWARGS:
                if k in s:
                    kwargs[k] = s[k]

            y_values = results.series(s['data'], smooth)
            if 'axis' in s and s['axis'] == 2:
                a = 1
            else:
                a = 0
            data[a] += y_values
            for r in settings.SCALE_DATA:
                data[a] += r.series(s['data'], smooth)
            config['axes'][a].plot(results.x_values,
                   y_values,
                   **kwargs)

        if 'scaling' in config:
            btm,top = config['scaling']
        else:
            btm,top = 0,100

        for a in range(len(config['axes'])):
            self._do_scaling(config['axes'][a], data[a], btm, top)

        self._do_legend(config)

    def _do_cdf_plot(self, results, config=None, axis=None):
        if axis is None:
            axis = self.plt.gca()
        if config is None:
            config = self.config

        data = []
        max_value = 0.0
        for s in config['series']:
            s_data = results.series(s['data'])
            if 'cutoff' in config:
                # cut off values from the beginning and end before doing the
                # plot; for e.g. pings that run long than the streams, we don't
                # want the unloaded ping values
                start,end = config['cutoff']
                s_data = s_data[int(start/settings.STEP_SIZE):-int(end/settings.STEP_SIZE)]
            d = sorted(filter(lambda x: x is not None, s_data))
            max_value = max([max_value]+d)
            data.append(d)

            for r in settings.SCALE_DATA:
                d_s = filter(lambda x: x is not None, r.series(s['data']))
                if d_s:
                    max_value = max([max_value]+d_s)


        x_values = list(frange(0, max_value, 0.1))

        for i,s in enumerate(config['series']):
            if not data[i]:
                continue
            kwargs = {}
            for k in PLOT_KWARGS:
                if k in s:
                    kwargs[k] = s[k]
            axis.plot(x_values,
                      [cum_prob(data[i], point) for point in x_values],
                      **kwargs)
        self._do_legend(config)


    def _do_meta_plot(self, results):
        for i,config in enumerate(self.configs):
            getattr(self, '_do_%s_plot' % config['type'])(results, config=config)

    def format(self, results):
        if not results:
            return

        getattr(self, '_do_%s_plot' % self.config['type'])(results)

        self._annotate_plot()

        # Since outputting image data to stdout does not make sense, we launch
        # the interactive matplotlib viewer if stdout is set for output.
        # Otherwise, the filename is passed to matplotlib, which selects an
        # appropriate output format based on the file name.
        if self.output == "-":
            self.plt.show()
        else:
            self.plt.savefig(self.output)


    def _annotate_plot(self):
        plot_title = settings.DESCRIPTION
        if 'description' in self.config:
            plot_title += "\n" + self.config['description']
        if settings.TITLE:
            plot_title += "\n" + settings.TITLE
        self.plt.suptitle(plot_title, fontsize=14)

        annotation_string = "Local/remote: %s/%s - Time: %s - Length/step: %ds/%.2fs" % (
            settings.LOCAL_HOST, settings.HOST,
            settings.TIME,
            settings.LENGTH, settings.STEP_SIZE)
        self.plt.suptitle(annotation_string,
                          x=0.5,
                          y=0.005,
                          horizontalalignment='center',
                          verticalalignment='bottom',
                          fontsize=8)

    def _do_legend(self, config):
        axes = config['axes']

        # Each axis has a set of handles/labels for the legend; combine them
        # into one list of handles/labels for displaying one legend that holds
        # all plot lines
        handles, labels = reduce(lambda x,y:(x[0]+y[0], x[1]+y[1]),
                                 [a.get_legend_handles_labels() for a in axes])

        # Shrink the current subplot by 20% in the horizontal direction, and
        # place the legend on the right of the plot.
        for a in axes:
            box = a.get_position()
            a.set_position([box.x0, box.y0, box.width * 0.8, box.height])


            kwargs = {}
            if 'legend_title' in config:
                kwargs['title'] = config['legend_title']

            a.legend(handles, labels,
                     bbox_to_anchor=(1.05, 1.0),
                     loc='upper left', borderaxespad=0.,
                     prop={'size':'small'},
                     **kwargs)

    def _do_scaling(self, axis, data, btm, top):
        """Scale the axis to the selected bottom/top percentile"""
        data = filter(lambda x: x is not None, data)
        top_percentile = self.np.percentile(data, top)*1.05
        btm_percentile = self.np.percentile(data, btm)*0.95
        axis.set_ylim(ymin=btm_percentile, ymax=top_percentile)
